sympy/polys/polytools.py

```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+
+        """
+        dom = f.get_domain()
+        if dom.is_FiniteField or not dom.is_Numerical:
+            raise DomainError("Cannot compute root bounds over %s" % dom)
+        if f.is_multivariate:
+            raise MultivariatePolynomialError(
+                "Must be a univariate polynomial")
+        if f.is_monomial:
+            raise PolynomialError("Monomial has no nonzero roots")
+
+        coeffs = f.all_coeffs()
+        while coeffs[-1] == 0:
+            coeffs.pop()
+        lc, tc = coeffs[0], coeffs[-1]
+        ub = 1 + max(abs(n / lc) for n in coeffs[1:])
+        lb_inv = 1 + max(abs(n / tc) for n in coeffs[:-1])
+
+        return 1/lb_inv, ub
+
+    def root_separation_lower_bound(f):
+        """
+        Compute a lower bound on the minimum distance between any two
+        roots of this polynomial.
+
+        Examples
+        ========
+
+        >>> from sympy import Poly
+        >>> from sympy.abc import x
+        >>> Poly(x**2 - 2).root_separation_lower_bound()
+        1
+        >>> (x * Poly(x**2 - 2)).root_separation_lower_bound()
+        12/17
+
+        Returns
+        =======
+
+        :py:class:`~.Expr`
+            A non-negative lower bound which will be positive if all the roots
+            are distinct, and rational if the coefficients of the polynomial
+            are rational.
+
+        Raises
+        ======
+
+        DomainError
+            If the domain of the polynomial is not a subring of the complex
+            numbers.
+        MultivariatePolynomialError
+            If the polynomial is not univariate.
+
+        """
+        dom = f.get_domain()
+        if dom.is_FiniteField or not dom.is_Numerical:
+            raise DomainError("Cannot compute root bounds over %s" % dom)
+        if f.is_multivariate:
+            raise MultivariatePolynomialError(
+                "Must be a univariate polynomial")
+        if f.is_monomial:
+            return Integer(0)
+
+        x = f.gen
+        y = Dummy('y')
+        g = f.subs(x, x + y)
+        h = Poly(resultant(f, g, gens=[x, y]), domain=f.domain)
+        # If f has roots ri, the roots of h are all the ri - rj.
+
+        if h.is_monomial:
+            # In this case all roots of f are the same.
+            return Integer(0)
+
+        lb, _ = h.root_bounds()
+
+        # Since h is an even function, can improve bound with square root.
+        if lb.is_rational:
+            # Preserve rational bound
+            return Rational(isqrt(lb.p*lb.q) + 1, lb.q)
+        else:
+            from sympy.functions.elementary.miscellaneous import sqrt
+            return sqrt(lb)
+
+    def root_comparison_tools(f):
+        """
+        Produce a threshold ``eps`` and custom evalf function ``evalf_`` such
+        that two roots ``a``, ``b`` of this polynomial must be equal if
+        ``abs(evalf_(a) - evalf_(b)) < eps``.
+
```

Comment:
I ran some tests, and

```python
A = evalf_(a)
for b in roots:
    abs(A - evalf_(b)) < eps
```

did not appear to be any faster than

```python
for b in roots:
    abs(evalf_(a) - evalf_(b)) < eps
```

(Due to caching, I suppose?)

So maybe this method can be simplified. It could be called `root_equality_test()`, and could simply return a binary function like

```python
lambda a, b: abs(evalf_(a) - evalf_(b)) < eps
```

(or one that incorporates the squared bound idea).

Reply:
A RootOf caches its internal interval so repeated calls to evalf are immediate:
```python
In [3]: r = RootOf(x**3 + x + 1, 0)

In [4]: r
Out[4]: 
       ⎛ 3           ⎞
CRootOf⎝x  + x + 1, 0⎠

In [5]: %time r.evalf()
CPU times: user 20 ms, sys: 0 ns, total: 20 ms
Wall time: 19 ms
Out[5]: -0.682327803828019

In [6]: %time r.evalf()
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 878 µs
Out[6]: -0.682327803828019
```
Internally `evalf` will try increasing the precision on subterms until it considers that the overall result has the accuracy it wants. So when you have `(a - b).evalf()` if `a = b` then `evalf` will increase the precision again and again until it hits the limit. Each time it does that RootOf needs to shrink its interval so we end up needing a much smaller interval then we would for `a.evalf()` or even `(a.evalf() - b).evalf()`. This is the part that is slowest.

These caches will affect your timings if you do everything in the same process. Instead make a complete Python script and time the whole operation:
```python
from sympy import *

x = Symbol('x')
p = x**4 + x**3 + x**2 + x + 1
rs = roots(p)
r3 = RootOf(p, 3)

def f():
    return [abs(r3 - r).evalf() for r in rs]

def g():
    r3e = r3.evalf()
    return [abs(r3e - r).evalf() for r in rs]

if True:
    f()
else:
    g()
```
That script will call `f` and the timing we get is:
```console
$ time python r.py

real	0m36.530s
user	0m36.496s
sys	0m0.024s
```
Now instead let's call `g()` (change the True to False):
```console
$ time python r.py

real	0m1.505s
user	0m1.464s
sys	0m0.040s
```
So `f` is a lot slower than `g` and this slowness is absolutely on a scale that is noticeable to a user.

If you look at the outputs of `f` and `g` you can also see that they don't get the same result:
```python
In [3]: f()
Out[3]: 
[1.90211303259031 - 0.e-25⋅ⅈ, 1.17557050458495 - 0.e-25⋅ⅈ, 1.90211303259031 + 0.e-22⋅ⅈ, 2.952700857
5875e-64 - 1.25078415499626e-63⋅ⅈ]

In [4]: g()
Out[4]: [1.90211303259031, 1.17557050458495, 1.90211303259031, 7.81510642993629e-9]
```
The absolute difference computed by `f` is smaller by 50 orders of magnitude.

Personally I think that both `f` and `g` are too slow. What needs to be improved for performance here is the algorithm that shrinks the intervals.

Reply:
Thanks. I didn't realize separate processes should be called in order to get a good comparison.

In the two forms I want to consider, the `evalf` always happens before the subtraction.
So, maybe the fact about caching already settles the question.
But just to be sure I add functions `h` and `k` to the script:

```python
import sys
from sympy import *

x = Symbol('x')
p = x**4 + x**3 + x**2 + x + 1
rs = roots(p)
r3 = RootOf(p, 3)

def f():
    return [abs(r3 - r).evalf() for r in rs]

def g():
    r3e = r3.evalf()
    return [abs(r3e - r).evalf() for r in rs]

def h():
    return [abs(r3.evalf() - r.evalf()) for r in rs]

def k():
    r3e = r3.evalf()
    return [abs(r3e - r.evalf()) for r in rs]

funcs = [f, g, h, k]

F = funcs[int(sys.argv[1])]
F()
```

and we get:

```bash
$ time python r.py 0

real    0m18.880s
user    0m18.381s
sys 0m0.153s

$ time python r.py 1

real    0m1.000s
user    0m0.937s
sys 0m0.056s

$ time python r.py 2

real    0m0.960s
user    0m0.891s
sys 0m0.052s

$ time python r.py 3

real    0m0.974s
user    0m0.895s
sys 0m0.057s
```

So I do think the `root_comparison_tools()` method I wrote could be simplified. I made it return the pair `(eps, evalf_)` because I imagined it might offer users an important performance improvement to let them use form `k()`. But, if in fact form `h()` is just as good, then it can be a simpler method that returns a single test function

```python
lambda a, b: abs(evalf_(a) - evalf_(b)) < eps
```

> Personally I think that both f and g are too slow. What needs to be improved for performance here is the algorithm that shrinks the intervals.

To me it depends on the application. For certain web apps, clicking and waiting one second to see a result might be just fine, whereas waiting 10s starts to feel unresponsive.